### PR TITLE
[iOS]Added SessionDetail UI

### PR DESCRIPTION
**Pull request by surajsau**
## Issue
- #380 

## Overview (Required)
- Added a few extension functions to `TimetableItem` 
  - `asSession()` and `asSpecial()` since I couldn't figure out an elegant way to handle type-casting for Kotlin sealed classes in Swift
  - <s>`durationInMinutes()` since I couldn't access `toComponent { .. }` in Swift</s>
  - <s>`durationString()` for convenience. Carried over the logic from `SessionScheduleInfo`</s>

### Remaining TODO
- [x] favorite button integration
- [x] adding actions for the various buttons (share, calendar etc.)
- [x] verify Speakers UI with fake data (in the mean time we wait for https://github.com/DroidKaigi/conference-app-2022/pull/388 to be merged)
- [x] cleanup

## Screenshot

https://user-images.githubusercontent.com/5493586/189830091-b902dfd6-79a3-47e3-8e3f-045179cd49ea.mp4

### CODE REVIEW

File: app-ios/Sources/SessionFeature/SessionView.swift [hunk 1]
```diff
@@ -1,39 +1,157 @@
...
 }
 
-public let sessionReducer = Reducer<SessionState, SessionAction, SessionEnvironment> { _, action, _ in
+public let sessionReducer = Reducer<SessionState, SessionAction, SessionEnvironment> { state, action, environment in
 
     switch action {
+    case .tapCalendar(let item):
+        return .run { @MainActor _ in
+            // TODO: Add event using EventKit
+        }
```
Comment (surajsau) [inline app-ios/Sources/SessionFeature/SessionView.swift:38]: I'm not 100% sure how to add event using EventKit. So instead I've added a TODO instead for future Issue.

File: app-ios/Sources/SessionFeature/SessionView.swift [hunk 2]
```diff
@@ -1,39 +1,157 @@
...
+                }
+                .padding()
+                .foregroundColor(AssetColors.onSurface.swiftUIColor)
+                .background(AssetColors.surface.swiftUIColor)
+            }
+            .foregroundColor(AssetColors.onBackground.swiftUIColor)
+            .background(AssetColors.background.swiftUIColor)
+            .sheet(isPresented: viewStore.binding(get: { $0.showShareSheet }, send: .nothing)) {
+                let text = "\(timeTableItem.title.currentLangTitle)\nhttps://droidkaigi.jp/2022/timetable/\(timeTableItem.id.value)"
+                ShareTextView(text: text)
```
Comment (surajsau) [inline app-ios/Sources/SessionFeature/SessionView.swift:152]: Is this the correct way to open a share sheet?

**Comment by surajsau** [comment]
Speaker view with fake data
スクリーンショット 2022-09-14 10 46 43

File: app-ios/Sources/SessionFeature/SessionTagsView.swift [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+                            if tag == tags.last! {
+                                height = 0 // last item
+                            }
+                            return result
+                        })
+                }
+            }
+        }
+    }
+
```
Comment (surajsau) [inline app-ios/Sources/SessionFeature/SessionTagsView.swift:43]: Implemented a custom view for overflowing tags. 
But it seems to have dimensions incorrectly set. When I remove it's bottom padding, the view below gets wrongly placed as follows:



Any hint would be of great help!

File: app-ios/Sources/SessionFeature/TagView.swift [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+        return Text(self.text)
+            .font(Font.system(size: 14, weight: .regular, design: .default))
+            .padding(.horizontal)
+            .padding(.vertical, 4)
+            .background(
+                Capsule()
+                    .fill(self.color)
+            )
+    }
+}
```
Comment (ry-itto) [inline app-ios/Sources/SessionFeature/TagView.swift:19]: How about use `CapsuleText` that is introduced in https://github.com/DroidKaigi/conference-app-2022/pull/402 🙏🏼 
But that view is defined in `TimetableFeature`, you can add that file copy in `SessionFeature`. If so, I'll define `CommonComponent` target and move those with another PR.

File: app-ios/Sources/SessionFeature/SessionView.swift [hunk 5]
```diff
@@ -1,39 +1,157 @@
...
+                    }
+                    .frame(width: 48, height: 48)
+                }
+                .padding()
+                .foregroundColor(AssetColors.onSurface.swiftUIColor)
+                .background(AssetColors.surface.swiftUIColor)
+            }
+            .foregroundColor(AssetColors.onBackground.swiftUIColor)
+            .background(AssetColors.background.swiftUIColor)
+            .sheet(isPresented: viewStore.binding(get: { $0.showShareSheet }, send: .nothing)) {
```
Comment (ry-itto) [inline app-ios/Sources/SessionFeature/SessionView.swift:150]: This send action should be hideShareSheet action.

e.g.
Change state to nil

File: app-ios/Sources/SessionFeature/SessionAssetsView.swift [hunk 6]
```diff
@@ -0,0 +1,42 @@
+import appioscombined
+import SwiftUI
+import Strings
```
Comment (ry-itto) [inline app-ios/Sources/SessionFeature/SessionAssetsView.swift:3]: Please add dependency about `Assets` , `Strings` and `Theme` to `Package.swift` 🙏🏼

File: app-ios/Sources/TimetableFeature/TimetableView.swift [hunk 7]
```diff
@@ -8,6 +8,7 @@ import Theme
 public struct TimetableState: Equatable {
     public var dayToTimetable: [DroidKaigi2022Day: Timetable]
     public var selectedDay: DroidKaigi2022Day
+    public var selectedItem: TimetableItemWithFavorite?
```
Comment (ry-itto) [inline app-ios/Sources/TimetableFeature/TimetableView.swift:11]: I think this state isn't needed 👀

File: app-ios/Sources/SessionFeature/SessionView.swift [hunk 8]
```diff
@@ -43,10 +161,9 @@ struct SessionView_Previews: PreviewProvider {
         SessionView(
             store: .init(
                 initialState: .init(
-                    timetableItem: Timetable.companion.fake().timetableItems.first!
-                ),
+                    timetableItemWithFavorite: TimetableItemWithFavorite(timetableItem: Timetable.companion.fake().timetableItems.first!, isFavorited: false)),
```
Comment (ry-itto) [inline app-ios/Sources/SessionFeature/SessionView.swift:164]: I think `fake()` should be useful 🙏🏼 
```suggestion
                    timetableItemWithFavorite: TimetableItemWithFavorite.companion.fake(),
```

File: app-ios/Sources/SessionFeature/SessionAssetsView.swift [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+    let asset: TimetableAsset
+    
+    public var body: some View {
+        VStack {
+            Text(L10n.Session.material)
+                .font(Font.system(size: 16, weight: .medium, design: .default))
+                .padding(.bottom)
+            if let videoUrl = self.asset.videoUrl {
+                HStack {
+                    Image(systemName: "video")
```
Comment (ry-itto) [inline app-ios/Sources/SessionFeature/SessionAssetsView.swift:17]: Please add asset and use it 🙏🏼

**Review by ry-itto — COMMENTED** [review]
Please fix lint violations 🙇🏼

File: app-ios/Sources/SessionFeature/SessionAssetsView.swift (hunk 9 above)
Comment (surajsau) [reply app-ios/Sources/SessionFeature/SessionAssetsView.swift:17]: Got it 👍

File: app-ios/Sources/SessionFeature/SessionDescriptionView.swift [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+            }
+        }
+    }
+}
+
+#if DEBUG
+struct SessionDescriptionView_Previews: PreviewProvider {
+    static var previews: some View {
+        SessionDescriptionView(
+            text: ""
```
[the remark this answers is not in the document]
Comment (surajsau) [reply app-ios/Sources/SessionFeature/SessionDescriptionView.swift:61]: I missed this part. Will add the fake data for previews 👍

File: app-ios/Sources/TimetableFeature/TimetableView.swift (hunk 7 above)
Comment (surajsau) [reply app-ios/Sources/TimetableFeature/TimetableView.swift:11]: Will remove it 🔥

File: app-ios/Sources/SessionFeature/SessionTagsView.swift (hunk 3 above)
Comment (surajsau) [reply app-ios/Sources/SessionFeature/SessionTagsView.swift:43]: Thanks for the hint! 👍

File: app-ios/Sources/SessionFeature/SessionDurationView.swift [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+    var durationView: some View {
+        HStack {
+            Image(systemName: "clock")
+                .padding(.trailing, 8)
+            Text(self.durationString(languageCode: Locale.current.languageCode))
+                .font(Font.system(size: 14, weight: .regular, design: .default))
+        }
+    }
+    
+    private func durationString(languageCode: String?) -> String {
```
[the remark this answers is not in the document]
Comment (surajsau) [reply app-ios/Sources/SessionFeature/SessionDurationView.swift:15]: Will use `DateIntervalFormatter` instead 👍

File: app-ios/Sources/SessionFeature/TagView.swift (hunk 4 above)
Comment (surajsau) [reply app-ios/Sources/SessionFeature/TagView.swift:19]: Sounds fine! For the meantime I'll copy it in Session 👍 CommonComponents in a future PR sounds good.

File: app-ios/Sources/SessionFeature/SessionTagsView.swift [hunk 12]
```diff
@@ -20,8 +59,31 @@ extension TimetableItem {
...
+            Color.clear
+                .preference(
+                    key: HeightPreferenceKey.self,
+                    value: geometry.frame(in: .local).size.height
+                )
+        }
+        .onPreferenceChange(HeightPreferenceKey.self) { height in
+            self.height = height
+        }
+    }
```
Comment (surajsau) [inline app-ios/Sources/SessionFeature/SessionTagsView.swift:96]: https://github.com/DroidKaigi/conference-app-2022/pull/403/commits/e50c77f6cbcde2bedaea1a7fda26728b9c7bc433#r970761357
> How about use [LazyHGrid](https://developer.apple.com/documentation/swiftui/lazyhgrid) with adaptive column?

I couldn't make it work properly with LazyGrid so instead tried determining the height of the TagsView and setting .frame(height:) to its parent Vstack.

File: app-ios/Sources/SessionFeature/SessionDescriptionView.swift (hunk 10 above)
[the remark this answers is not in the document]
Comment (surajsau) [reply app-ios/Sources/SessionFeature/SessionDescriptionView.swift]: Fixed and added previews for other Views also in https://github.com/DroidKaigi/conference-app-2022/pull/403/commits/d29010c8d34f65941cda35e01a00b7fcca4fcef3

File: app-ios/Sources/SessionFeature/SessionView.swift (hunk 8 above)
Comment (surajsau) [reply app-ios/Sources/SessionFeature/SessionView.swift]: Fixed and added previews for other Views also in https://github.com/DroidKaigi/conference-app-2022/commit/d29010c8d34f65941cda35e01a00b7fcca4fcef3

File: app-ios/Sources/TimetableFeature/TimetableView.swift (hunk 7 above)
Comment (surajsau) [reply app-ios/Sources/TimetableFeature/TimetableView.swift:11]: Correct me if I'm wrong, but I guess this is required for opening `SessionView`.

```Swift
// AppView.swift 167-170
case let .timetable(.selectItem(item)):
    state.sessionState = .init(
        timetableItemWithFavorite: item
)
```

File: app-ios/Sources/SessionFeature/SessionAssetsView.swift (hunk 6 above)
Comment (surajsau) [reply app-ios/Sources/SessionFeature/SessionAssetsView.swift:4]: Done in  https://github.com/DroidKaigi/conference-app-2022/commit/cc4616edea505764c7b5ae921b873d0a817c1132

File: app-ios/Sources/SessionFeature/SessionView.swift [hunk 13]
```diff
@@ -1,39 +1,157 @@
...
+                    }
+                    .padding(.horizontal)
+                }
+
+                HStack {
+                    HStack {
+                        Button {
+                            viewStore.send(.tapShare)
+                        } label: {
+                            Image(systemName: "square.and.arrow.up")
```
[the remark this answers is not in the document]
Comment (surajsau) [reply app-ios/Sources/SessionFeature/SessionView.swift]: Added all the assets and replaced Image with them in https://github.com/DroidKaigi/conference-app-2022/commit/cc4616edea505764c7b5ae921b873d0a817c1132

File: app-ios/Sources/SessionFeature/SessionView.swift [hunk 14]
```diff
@@ -1,39 +1,157 @@
...
+                    Image(systemName: "xmark")
+                }
+                .padding(.leading)
+                .padding([.top, .bottom], 16)
+
+                ScrollView {
+                    VStack(alignment: .leading) {
+                        Text(timeTableItem.title.currentLangTitle)
+                            .frame(maxWidth: .infinity, alignment: .topLeading)
+                            .font(Font.system(size: 24, weight: .medium, design: .default))
```
[the remark this answers is not in the document]
Comment (surajsau) [reply app-ios/Sources/SessionFeature/SessionView.swift]: Fixed sizes and dimesions as per figma in https://github.com/DroidKaigi/conference-app-2022/pull/403/commits/cac96d1b222f6ecb50cac272f3a2d96a684b84c5

File: app-ios/Sources/SessionFeature/SessionView.swift [hunk 15]
```diff
@@ -1,39 +1,157 @@
...
     private let store: Store<SessionState, SessionAction>
 
     public init(store: Store<SessionState, SessionAction>) {
         self.store = store
     }
 
     public var body: some View {
-        Text("")
+        WithViewStore(store) { viewStore in
+            let timeTableItem = viewStore.timetableItemWithFavorite.timetableItem
```
[the remark this answers is not in the document]
Comment (surajsau) [reply app-ios/Sources/SessionFeature/SessionView.swift]: Done 👍

File: app-ios/Sources/SessionFeature/TagView.swift (hunk 4 above)
Comment (surajsau) [reply app-ios/Sources/SessionFeature/TagView.swift]: Added a  copy of `CapsuleText` in SessionFeature 👍

File: app-ios/Sources/SessionFeature/SessionView.swift [hunk 16]
```diff
@@ -1,39 +1,157 @@
...
 
     switch action {
+    case .tapCalendar(let item):
+        return .run { @MainActor _ in
+            // TODO: Add event using EventKit
+        }
+        .receive(on: DispatchQueue.main.eraseToAnyScheduler())
+        .eraseToEffect()
+    case .tapFavorite(let sessionId, let isFavorite):
+        state.timetableItemWithFavorite = TimetableItemWithFavorite(timetableItem: state.timetableItemWithFavorite.timetableItem, isFavorited: !isFavorite)
```
[the remark this answers is not in the document]
Comment (surajsau) [reply app-ios/Sources/SessionFeature/SessionView.swift]: Done 👍

File: app-ios/Sources/SessionFeature/SessionSpeakersView.swift [hunk 17]
```diff
@@ -0,0 +1,36 @@
+import appioscombined
+import Strings
+import SwiftUI
+import Theme
+
+struct SessionSpeakersView: View {
+
+    let speakers: [TimetableSpeaker]
+    
+    public var body: some View {
```
[the remark this answers is not in the document]
Comment (surajsau) [reply app-ios/Sources/SessionFeature/SessionSpeakersView.swift]: Done 👍

File: app-ios/Sources/SessionFeature/SessionDurationView.swift (hunk 11 above)
[the remark this answers is not in the document]
Comment (surajsau) [reply app-ios/Sources/SessionFeature/SessionDurationView.swift]: Using `DateIntervalFormatter` it seems to be slightly different format in Japanese. Will this be fine ?
スクリーンショット 2022-09-15 22 40 28

File: app-ios/Sources/SessionFeature/SessionView.swift (hunk 2 above)
Comment (surajsau) [reply app-ios/Sources/SessionFeature/SessionView.swift]: Sorry I didn't completely understand how to implement your idea.

**Comment by surajsau** [comment]
Don't know why the build failed 🤔 corrected the Swiflint issues too in [89c0423](https://github.com/DroidKaigi/conference-app-2022/pull/403/commits/89c0423657bb107f4327c7f61723293d4d4df965)

This is how the screen looks after fixing bugs as directed.

https://user-images.githubusercontent.com/5493586/190535139-09f08413-ec11-44b1-a7f0-95654b9aefe8.mp4

File: app-ios/Sources/SessionFeature/SessionDurationView.swift (hunk 11 above)
[the remark this answers is not in the document]
Comment (ry-itto) [reply app-ios/Sources/SessionFeature/SessionDurationView.swift]: It's fine!

File: app-ios/Sources/SessionFeature/SessionTagsView.swift (hunk 12 above)
Comment (ry-itto) [reply app-ios/Sources/SessionFeature/SessionTagsView.swift:96]: Okey!
I'll try to implement with `LazyGrid` in other issues and PR! This is ok for now 👍🏼

File: app-ios/Sources/TimetableFeature/TimetableView.swift (hunk 7 above)
Comment (surajsau) [reply app-ios/Sources/TimetableFeature/TimetableView.swift]: You were right! Fixed it in https://github.com/DroidKaigi/conference-app-2022/pull/403/commits/becf16344c9fca6c420c0118858ad89980122209

**Comment by surajsau** [comment]
I'm unable to undertstand why the build failed ([Details](https://github.com/DroidKaigi/conference-app-2022/actions/runs/3075373055/jobs/4968755703)) 😞

File: app-ios/Sources/TimetableFeature/TimetableView.swift [hunk 18]
```diff
@@ -63,8 +63,6 @@ public let timetableReducer = Reducer<TimetableState, TimetableAction, Timetable
     case let .selectDay(day):
         state.selectedDay = day
         return .init(value: .refresh)
-    case .selectItem:
-        return .none
```
Comment (woxtu) [inline app-ios/Sources/TimetableFeature/TimetableView.swift:67]: Maybe removing this case gets the build error.

Comment (ry-itto) [reply app-ios/Sources/TimetableFeature/TimetableView.swift:67]: Exactly. 

@surajsau 
Please revert this change 🙏🏼

Comment (surajsau) [reply app-ios/Sources/TimetableFeature/TimetableView.swift:67]: Done and fixed 👍

**Review by ry-itto — APPROVED** [review]
LGTM! Thank you for contributing!

**Pull request merged**